File: common/src/main/java/com/andyanika/translator/common/interfaces/LocalRepository.java
package com.andyanika.translator.common.interfaces;

import com.andyanika.translator.common.models.FavoriteModel;
import com.andyanika.translator.common.models.LanguageCode;
import com.andyanika.translator.common.models.TranslateDirection;
import com.andyanika.translator.common.models.TranslateRequest;
import com.andyanika.translator.common.models.TranslateResult;

import java.util.List;

import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

public interface LocalRepository {
    Single<TranslateResult> translate(TranslateRequest request);
    Single<TranslateResult> addTranslation(TranslateResult translateResult);

    Flowable<List<FavoriteModel>> getHistory();

    Flowable<List<FavoriteModel>> getFavorites();
    void addFavorites(int wordId);
    void removeFavorite(int wordId);

    void setLanguageDirection(TranslateDirection<LanguageCode> direction);
    Observable<LanguageCode> getAvailableLanguages();
    Observable<LanguageCode> getSrcLanguage();
    Observable<LanguageCode> getDstLanguage();
}

File: common/src/main/java/com/andyanika/translator/common/interfaces/LocalRepository.kt
package com.andyanika.translator.common.interfaces

import com.andyanika.translator.common.models.*
import kotlinx.coroutines.flow.Flow

interface LocalRepository {
    suspend fun translate(request: TranslateRequest): TranslateResult?
    suspend fun addTranslation(translateResult: TranslateResult)
    fun getHistory(): Flow<List<FavoriteModel>>
    fun getFavorites(): Flow<List<FavoriteModel>>
    suspend fun addFavorites(wordId: Int)
    suspend fun removeFavorite(wordId: Int)
    suspend fun setLanguageDirection(direction: TranslateDirection<LanguageCode>)
    suspend fun getAvailableLanguages(): List<LanguageCode>

    fun observeSrcLanguage(): Flow<LanguageCode>
    fun observeDstLanguage(): Flow<LanguageCode>
    suspend fun getSrcLanguage(): LanguageCode
    suspend fun getDstLanguage(): LanguageCode
}
